## Universidad Politécnica de San Luis Potosí
###  Regresión Lineal Simple

En esta práctica aplicaremos el algoritmo de Regresión Lineal Simple

sirve para pronosticar un valor numérico

Este algoritmo funciona con una variable independiente y una variable dependiente, ambas deben ser numéricas

In [ ]:
#%pip install pandas
#%pip install seaborn
#%pip install scikit-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.5/12.5 MB 151.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.1/35.1 MB 114.2 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3/3 [scikit-learn] [scikit-learn]
Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as ps

pandas: Funciones para importar, manipular y realizar análisis de datos

In [4]:
import numpy as np

numpy:  Biblioteca para trabajar con matrices de n-dimensiones

In [7]:
import seaborn as sb

seaborn: Herramienta para gráficos

In [8]:
import matplotlib.pyplot as pl

matplotlib: Gráficas en 2D y 3D

In [13]:
from sklearn.metrics import classification_report
# esta línea trae la función para generar un reporte de la clasificación
from sklearn import linear_model
# esta línea es la que carga la función que calcula la regresión
from sklearn.metrics import mean_squared_error, r2_score
# son funciones para calcular medidas de eficiencia del modelo de regresión
# error cuadrado medio y r cuadrada


In [ ]:
from google.colab import drive
drive.mount('/content/drive')
#montamos Drive

Mounted at /content/drive


In [14]:
# cargamos los datos
# datos = ps.read_csv("/content/drive/MyDrive/copiadoras.csv")
datos =  ps.read_csv("./copiadoras.csv")

In [15]:
datos

Vendedor  Llamadas  Ventas
0         1        20      30
1         2        40      60
2         3        20      40
3         4        30      60
4         5        10      30
5         6        10      40
6         7        20      40
7         8        20      50
8         9        20      30
9        10        30      70

In [16]:
# Calculamos la correlación
correlacion = datos.corr(method='pearson')
correlacion

Vendedor  Llamadas    Ventas
Vendedor  1.000000 -0.159745  0.191977
Llamadas -0.159745  1.000000  0.759014
Ventas    0.191977  0.759014  1.000000

Nota importante: No tomar en cuenta la variable Vendedor, es una variable que solmente sirve para identificar al vendedor.

Observamos que Ventas y Llamadas están relacionadas ya que la correlación es de 0.759, entonces si podemos calcular un modelo de Regresión.

(Recordar que el valor debe ser estar entre -0.4 y -1  o entre 0.4 y 1)

In [17]:
# Dejamos la variable independiente en la variable X
# Ojo: No debemos tomar en cuenta el número del vendedor
X = datos['Llamadas']
X

0    20
1    40
2    20
3    30
4    10
5    10
6    20
7    20
8    20
9    30
Name: Llamadas, dtype: int64

In [18]:
# Dejamos la variable dependiente en la variable y
y = datos['Ventas']
y

0    30
1    60
2    40
3    60
4    30
5    40
6    40
7    50
8    30
9    70
Name: Ventas, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
# Esta función nos permitirá separar el conjunto de datos en dos, uno de entrenamiento
# y el otro de prueba
# Así funciona el Aprendizaje de Máquina Supervisado


In [ ]:
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Con esta instrucción dividimos los datos en cuatro conjuntos:
# X_train, y_train  datos de entrenamiento
# X_test,  y_test   datos de prueba
# el de entrenamiento contiene el 70% de los datos
# el de prueba tiene el 30% de los datos (por eso el valor test_size=0.3)
# random_state es para que los seleccione de manera aleatorria, 42 en la semila de aleatorización, puede ser cualquier número

In [ ]:
X_train # vemos cómo quedó

0    20
7    20
2    20
9    30
4    10
3    30
6    20
Name: Llamadas, dtype: int64

In [ ]:
y_train   # vemos cómo quedó

0    30
7    50
2    40
9    70
4    30
3    60
6    40
Name: Ventas, dtype: int64

In [ ]:
X_test  # vemos cómo quedó

8    20
1    40
5    10
Name: Llamadas, dtype: int64

In [ ]:
y_test  # vemos cómo quedó

8    30
1    60
5    40
Name: Ventas, dtype: int64

In [ ]:
regl = linear_model.LinearRegression()
# Hacemos la instancia para la función de regresión lineal

In [ ]:
X_train = np.array(X_train)
# lo convertimos a array para que en la línea siguiente redimensionar con reshape
# hacemos este paso para redimensionar el array, porque solo tenemos una
# variable

In [ ]:
regl.fit(X_train.reshape(-1, 1), y_train)


LinearRegression()

regl.fit  es la instrucción para realizar la regresión, recordar de las líneas anteriores que regl es la instancia de linear_model.LinearRegression

Con la línea anterior hemos hecho el ajuste del modelo de regresión a los 7 datos de entrenamiento

In [ ]:
X_test = np.array(X_test)

In [ ]:
# Ya hemos entrenado u ajustado el modelo de regresión con los datos de entrenamiento,
# ahora vamos a introducir los datos de prueba que están en X_test
# para realizar la estimación (pronóstico) de las ventas
pronostico = regl.predict(X_test.reshape(-1, 1))

In [ ]:
pronostico
# los valores obtenidos fueron: 43, 81 y 24
# los valores orginales eran 30, 60 y 40  (los que están en y_test)
# para el primer valor se equivocó en 13 unidades (43-30)
# NO HAY modelos de Regresión que sean exactos, SIEMPRE habrá errores en la predicción
# lo IMPORTANTE es tratar que es error sea mínimo

array([43., 81., 24.])

In [ ]:
print('Coeficiente: ', regl.coef_)

Coeficiente:  [1.9]


In [ ]:
print('Término independiente: ', regl.intercept_)

Término independiente:  5.000000000000007


La ecuación de regresión nos queda así:

y = 5 + 1.9 * X

y = 5 + 1.9 * 20(llamadas)

y = 43

In [ ]:
print ('Error cuadrado medio: ', mean_squared_error(y_test, pronostico))

Error cuadrado medio:  288.6666666666667


Si tuvieramos varios modelos, el mejor es el que tiene el menor error cuadrado medio.


 Este tipo de modelos de pronóstico los utilizamos para calcular valores que no conocemos, por ejemplo:

 queremos saber cuál sería el número de copiadoras vendidas si se hacen 25 llamadas ó 35 llamadas.



In [ ]:
valores_nuevos = np.array([25,35])

In [ ]:
pronostico_nuevos = regl.predict(valores_nuevos.reshape(-1, 1))
pronostico_nuevos

array([52.5, 71.5])

Con 25 llamadas se estima que se venderían 52 copiadoras y con 35 llamadas se venderían 71 copiadoras

Fin de la práctica